src/fourier.jl

```diff
@@ -0,0 +1,781 @@
+    new_coherence = 1.0
+    old_coherence = 0.0
+    count = 0
+    while (!(≈(new_coherence, old_coherence, atol=0.01)) && count< 40)
+        old_coherence = new_coherence
+        bsq = bias_term(power1, power2, power1_noise, power2_noise,
+                        n_ave, intrinsic_coherence=new_coherence)
+        den = (power1 - power1_noise) * (power2 - power2_noise)
+        num = real(cross_power * conj(cross_power)) - bsq
+        if num < 0
+            num = real(cross_power * conj(cross_power))
+        end
+        new_coherence = num / den
+        count += 1
+    end
+    return new_coherence                                       
+end
+
+estimate_intrinsic_coherence(cross_power:: AbstractVector{<:Complex}, 
+                             power1:: AbstractVector{<:Real}, power2:: AbstractVector{<:Real}, 
+                             power1_noise::Real, power2_noise::Real, n_ave::Integer)= 
+                                _estimate_intrinsic_coherence_single.(
+                                    cross_power, power1, power2, 
+                                    power1_noise, power2_noise, n_ave)
+
+function error_on_averaged_cross_spectrum(cross_power:: AbstractVector{<:Complex}, 
+                                          seg_power:: AbstractVector{<:Real}, 
+                                          ref_power:: AbstractVector{<:Real}, n_ave::Integer,
+                                          seg_power_noise::Real, ref_power_noise::Real;
+                                          common_ref::Bool=false)
+    two_n_ave = 2 * n_ave
+    if common_ref
+        Gsq = (cross_power * conj(cross_power)).real
+        bsq = bias_term(seg_power, ref_power, seg_power_noise, ref_power_noise,
+                        n_ave)
+        frac = (Gsq - bsq) / (ref_power - ref_power_noise)
+        power_over_2n = ref_power / two_n_ave
+
+        # Eq. 18
+        dRe = dIm = dG = sqrt(power_over_2n * (seg_power - frac))
+        # Eq. 19
+        dphi = sqrt(power_over_2n * (seg_power / (Gsq - bsq) -
+                       1 / (ref_power - ref_power_noise)))
+
+    else
+        PrPs = ref_power * seg_power
+        dRe = sqrt((PrPs + cross_power.real ^ 2 - cross_power.imag ^ 2) /
+                      two_n_ave)
+        dIm = sqrt((PrPs - cross_power.real ^ 2 + cross_power.imag ^ 2) /
+                      two_n_ave)
+        gsq = raw_coherence(cross_power, seg_power, ref_power,
+                            seg_power_noise, ref_power_noise, n_ave)
+        dphi = sqrt((1 - gsq) / (2 * gsq * n_ave))
+        dG = sqrt(PrPs / n_ave)
+    end
+
+    return dRe, dIm, dphi, dG
+end
+
+function cross_to_covariance(cross_power:: AbstractVector{<:Complex}, 
+                             ref_power:: AbstractVector{<:Real}, 
+                             ref_power_noise::Real, 
+                             delta_nu::T) where {T<:Union{AbstractVector{<:Real},Real}}
+    return cross_power .* sqrt.(delta_nu ./ (ref_power .- ref_power_noise))
+end
```

Comment:
Once again, the pattern in Julia is to define a function `f(x::Real) = ...` only on scalars, and then call it `f.(V)` where `V::AbstractVector{<:Real}`.  There is no need to define the function on scalars _and_ vectors, it's a useless and confusing duplicate.

Reply:
Ok but I want first two arguments to always be array, should I use Ref there?